source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -794,45 +796,74 @@ pub trait EnvoyHttpFilter {
   /// // Mutable static storage is used for the test to simulate the response body operation.
   /// static mut BUFFER: [u8; 10] = *b"helloworld";
   /// envoy_filter
-  ///   .expect_get_request_body()
+  ///   .expect_get_received_request_body()
   ///   .returning(|| Some(vec![EnvoyMutBuffer::new(unsafe { &mut BUFFER })]));
-  /// envoy_filter.expect_drain_request_body().return_const(true);
+  /// envoy_filter
+  ///   .expect_drain_received_request_body()
+  ///   .return_const(true);
   ///
   ///
-  /// // Calculate the size of the request body in bytes.
-  /// let buffers = envoy_filter.get_request_body().unwrap();
+  /// // Calculate the size of the new received request body in bytes.
+  /// let buffers = envoy_filter.get_received_request_body().unwrap();
   /// let mut size = 0;
   /// for buffer in &buffers {
   ///   size += buffer.as_slice().len();
   /// }
   /// assert_eq!(size, 10);
   ///
-  /// // drain the entire request body.
-  /// assert!(envoy_filter.drain_request_body(10));
+  /// // drain the new received request body.
+  /// assert!(envoy_filter.drain_received_request_body(10));
   ///
   /// // Now start writing new data from the beginning of the request body.
   /// ```
   ///
   /// This returns None if the request body is not available.
-  fn get_request_body<'a>(&'a mut self) -> Option<Vec<EnvoyMutBuffer<'a>>>;
+  fn get_received_request_body<'a>(&'a mut self) -> Option<Vec<EnvoyMutBuffer<'a>>>;
+
+  /// Similar to `get_received_request_body`, but returns the historically buffered request body.
+  fn get_buffered_request_body<'a>(&'a mut self) -> Option<Vec<EnvoyMutBuffer<'a>>>;
 
-  /// Drain the given number of bytes from the front of the request body.
+  /// Drain the given number of bytes from the front of the new received request body.
+  /// This should only be used in the `on_request_body` callback.
   ///
   /// Returns false if the request body is not available.
   ///
   /// Note that after changing the request body, it is caller's responsibility to modify the
   /// content-length header if necessary.
-  fn drain_request_body(&mut self, number_of_bytes: usize) -> bool;
+  fn drain_received_request_body(&mut self, number_of_bytes: usize) -> bool;
+
+  /// Similar to `drain_received_request_body`, but drains from the historically buffered request
+  /// body. This should only be called by the latest data processing filter in the filter chain.
+  /// That is say the filters after this filter in the filter chain should have not accessed the
+  /// request body yet.
```

Comment:
Sorry I am not following this.. are you saying that any subsequent filters should not access body buffer once this is used, even when they are reading, not modifying? Could you elaborate on why?